### ISSUE TITLE
Unknown SSL protocol error in connection to github.com:443

### ISSUE THREAD

**Issue by Andrey-Bushman** [git-for-windows/git#234]
Windows 7 x64. Computer are located in a domain of company. Installed Git-2.4.5.1-4th-release-candidate-64-bit.exe. I get an error, when I try to get repository from GitHub:

Developer@BUSHCOMP MINGW64 /d/_git.sandbox
$ git clone https://github.com/progit/progit2-ru.git
Клонирование в «progit2-ru»…
fatal: unable to access 'https://github.com/progit/progit2-ru.git/': Unknown SSL protocol error in connection to github.com:443

How can I solve it?

**Comment by kostix** [comment]
Please try with `GIT_CURL_VERBOSE=1` in the environment (`LANG=C` would also be a bonus).

If you're using Git Bash, then do

    export GIT_CURL_VERBOSE=1
    export LANG=C
    git clone ...

and if you're using regular `cmd.exe`, do

    set GIT_CURL_VERBOSE=1

...and so on.

This would make the HTTP[S] client library Git uses be chatty about what it does.

**Comment by Andrey-Bushman** [comment]
In the .gitconfig I added connection info:

[http]
    proxy = http://Developer:my_password@proxy2:8080

[https]
    proxy = https://Developer:my_password@proxy2:8080

I got this:

Developer@BUSHCOMP MINGW64 /d/_git.sandbox
$ export GIT_CURL_VERBOSE=1

Developer@BUSHCOMP MINGW64 /d/_git.sandbox
$ export LANG=C

Developer@BUSHCOMP MINGW64 /d/_git.sandbox
$ git clone https://github.com/nunit/nunit.git
Cloning into 'nunit'...
* Couldn't find host github.com in the _netrc file; using defaults
* timeout on name lookup is not supported
*   Trying 192.168.123.12...
* Connected to proxy2 (192.168.123.12) port 8080 (#0)
* Establish HTTP proxy tunnel to github.com:443
> CONNECT github.com:443 HTTP/1.1
Host: github.com:443
User-Agent: git/2.4.5.windows.1
Proxy-Connection: Keep-Alive

< HTTP/1.1 407 Proxy Authentication Required ( Для выполнения запроса компоненту Forefront TMG требуется авторизация. Доступ к фильтру веб-прокси запрещен.  )
< Via: 1.1 PROXY2
< Proxy-Authenticate: Negotiate
< Proxy-Authenticate: Kerberos
< Proxy-Authenticate: NTLM
< Connection: close
< Proxy-Connection: close
< Pragma: no-cache
< Cache-Control: no-cache
< Content-Type: text/html
< Content-Length: 1033
<
* Ignore 1033 bytes of response-body
* Connect me again please
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: C:/Users/developer/AppData/Local/Programs/Git/mingw64/ssl/certs/ca-bundle.crt
  CApath: none
* Unknown SSL protocol error in connection to github.com:443
* Closing connection 0
* Couldn't find host github.com in the _netrc file; using defaults
* timeout on name lookup is not supported
* Hostname proxy2 was found in DNS cache
*   Trying 192.168.123.12...
* Connected to proxy2 (192.168.123.12) port 8080 (#1)
* Establish HTTP proxy tunnel to github.com:443
> CONNECT github.com:443 HTTP/1.1
Host: github.com:443
User-Agent: git/2.4.5.windows.1
Proxy-Connection: Keep-Alive

< HTTP/1.1 407 Proxy Authentication Required ( Для выполнения запроса компоненту Forefront TMG требуется авторизация. Доступ к фильтру веб-прокси запрещен.  )
< Via: 1.1 PROXY2
< Proxy-Authenticate: Negotiate
< Proxy-Authenticate: Kerberos
< Proxy-Authenticate: NTLM
< Connection: close
< Proxy-Connection: close
< Pragma: no-cache
< Cache-Control: no-cache
< Content-Type: text/html
< Content-Length: 1033
<
* Ignore 1033 bytes of response-body
* Connect me again please
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: C:/Users/developer/AppData/Local/Programs/Git/mingw64/ssl/certs/ca-bundle.crt
  CApath: none
* Unknown SSL protocol error in connection to github.com:443
* Closing connection 1
* Couldn't find host github.com in the _netrc file; using defaults
* timeout on name lookup is not supported
* Hostname proxy2 was found in DNS cache
*   Trying 192.168.123.12...
* Connected to proxy2 (192.168.123.12) port 8080 (#2)
* Establish HTTP proxy tunnel to github.com:443
* Proxy auth using Negotiate with user 'Developer'
> CONNECT github.com:443 HTTP/1.1
Host: github.com:443
Proxy-Authorization: Negotiate TlRMTVNTUAABAAAAt4II4gAAAAAAAAAAAAAAAAAAAAAGAbEdAAAADw==
User-Agent: git/2.4.5.windows.1
Proxy-Connection: Keep-Alive

< HTTP/1.1 407 Proxy Authentication Required ( Access is denied.  )
< Via: 1.1 PROXY2
< Proxy-Authenticate: Negotiate TlRMTVNTUAACAAAAFAAUADgAAAA1gonikb+EQSfPwI0AAAAAAAAAAIIAggBMAAAABgGwHQAAAA9IAFkAUABSAE8AUwBUAFIATwBZAAIAFABIAFkAUABSAE8AUwBUAFIATwBZAAEADABQAFIATwBYAFkAMgAEABAAaAB5AHAAcgBvAHMAdAByAAMAHgBwAHIAbwB4AHkAMgAuAGgAeQBwAHIAbwBzAHQAcgAFABAAaAB5AHAAcgBvAHMAdAByAAcACABk9uqlPLrQAQAAAAA=
< Connection: Keep-Alive
< Proxy-Connection: Keep-Alive
< Pragma: no-cache
< Cache-Control: no-cache
< Content-Type: text/html
< Content-Length: 0
<
* TUNNEL_STATE switched to: 0
* Establish HTTP proxy tunnel to github.com:443
* Proxy auth using Negotiate with user 'Developer'
> CONNECT github.com:443 HTTP/1.1
Host: github.com:443
Proxy-Authorization: Negotiate TlRMTVNTUAADAAAAGAAYAHoAAAAgASABkgAAAAAAAABYAAAAEgASAFgAAAAQABAAagAAABAAEACyAQAANYKI4gYBsR0AAAAPtULiFSvTc8WASGK+MMXv1UQAZQB2AGUAbABvAHAAZQByAEIAVQBTAEgAQwBPAE0AUAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA1V5pwb7BOMg+LNkyXss1XAQEAAAAAAABk9uqlPLrQATMvpYJute4gAAAAAAIAFABIAFkAUABSAE8AUwBUAFIATwBZAAEADABQAFIATwBYAFkAMgAEABAAaAB5AHAAcgBvAHMAdAByAAMAHgBwAHIAbwB4AHkAMgAuAGgAeQBwAHIAbwBzAHQAcgAFABAAaAB5AHAAcgBvAHMAdAByAAcACABk9uqlPLrQAQYABAACAAAACAAwADAAAAAAAAAAAAAAAAAgAABTOYTKmo0But0uuVbnX8udmrGxdPuYoUK79IKjJqXGPwoAEAAAAAAAAAAAAAAAAAAAAAAACQAWAEgAVABUAFAALwBwAHIAbwB4AHkAMgAAAAAAAAAAAAAAAADX0VMoRH4oPjQlqesPxr8I
User-Agent: git/2.4.5.windows.1
Proxy-Connection: Keep-Alive

< HTTP/1.1 407 Proxy Authentication Required ( Для выполнения запроса компоненту Forefront TMG требуется авторизация. Доступ к фильтру веб-прокси запрещен.  )
< Via: 1.1 PROXY2
< Proxy-Authenticate: Negotiate
< Proxy-Authenticate: Kerberos
< Proxy-Authenticate: NTLM
< Connection: close
< Proxy-Connection: close
< Pragma: no-cache
< Cache-Control: no-cache
< Content-Type: text/html
< Content-Length: 1033
<
* Received HTTP code 407 from proxy after CONNECT
* Closing connection 2
fatal: unable to access 'https://github.com/nunit/nunit.git/': Unknown SSL protocol error in connection to github.com:443

**Comment by Andrey-Bushman** [comment]
When I used version 1.9.5 from here: https://git-scm.com/download/win it worked fine. :(((

**Comment by dscho** [comment]
Looks to me as if the proxy talked an SSL protocol that is no longer supported. I remember that one of the older SSL protocols was disabled recently because it poses security concerns. Maybe that's it?

**Comment by kostix** [comment]
@dscho, I disagree: it's impossible to proxy TLS connections because to do this, you'd have to actually MitM it (on a side note -- you might find [this](http://arstechnica.com/security/2015/02/lenovo-pcs-ship-with-man-in-the-middle-adware-that-breaks-https-connections/) amusing to read). Hence the client does not use `GET` or `POST` (or other HTTP verb) but rather issues `CONNECT` to the specified host and port.  The proxy is there to handle authentication, if needed, and when all is OK, it merely creates a tunnel for the client and basically shovels opaque data between the server and the client. The port on which proxy listens is itself not TLS-enabled in any way.

But I'm afraid the root cause is that the proxy never actually answered 200 OK to the client: no matter what it sent, it kept sending 407 back.

**Comment by kostix** [comment]
Andrey, one more thing: is this possible to negotiate with your admins to let direct connections to github's 443 for some time?  You could then try to factor out the proxy by directly testing whether Git's OpenSSL is able to connect using something like

    openssl s_client -host github.com -port 443

If it connects OK, the problem is likely rooted in the way libCURL talks with the proxy.

Or may be in the settings it does to the OpenSSL's context (see my previous comment).

**Comment by kostix** [comment]
OK, tested using `curl` directly (it's also bundled with GfW) through my corporate proxy (some old Squid instance which supports NTLM against Windows AD) -- and it worked OK:

````
kostix@programmer MINGW32 /c/tmp
$ curl -x 192.168.2.20:8080 --proxy-ntlm --proxy-user : --verbose https://githu
b.com/nunit/nunit.git/info/refs?service=git-upload-pack
* timeout on name lookup is not supported
*   Trying 192.168.2.20...
* Connected to 192.168.2.20 (192.168.2.20) port 8080 (#0)
* Establish HTTP proxy tunnel to github.com:443
* Proxy auth using NTLM with user ''
> CONNECT github.com:443 HTTP/1.1
> Host: github.com:443
> Proxy-Authorization: NTLM TlRMTVNTUAABAAAAB7IIogkACQAyAAAACgAKACgAAAAFASgKAAAA
D1BST0dSQU1NRVJET01BSU4wMDc=
> User-Agent: curl/7.43.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.0 407 Proxy Authentication Required
< Server: squid/2.7.STABLE3
< Date: Thu, 09 Jul 2015 17:08:35 GMT
< Content-Type: text/html
< Content-Length: 1274
< Expires: Thu, 09 Jul 2015 17:08:35 GMT
< X-Squid-Error: ERR_CACHE_ACCESS_DENIED 0
< Proxy-Authenticate: NTLM TlRMTVNTUAACAAAAEgASADAAAAAFgomizq2FDxfujTgAAAAAAAAAA
HwAfABCAAAARABPAE0AQQBJAE4AMAAwADcAAgASAEQATwBNAEEASQBOADAAMAA3AAEAEABQAFIATwBYA
FkAUwBSAFYABAAaAGQAbwBtAGEAaQBuADAAMAA3AC4AYwBvAG0AAwAsAHAAcgBvAHgAeQBzAHIAdgAuA
GQAbwBtAGEAaQBuADAAMAA3AC4AYwBvAG0AAAAAAA==
< X-Cache: MISS from proxy.domain.local
< X-Cache-Lookup: NONE from proxy.domain.local:8080
< Via: 1.0 proxy.domain.local:8080 (squid/2.7.STABLE3)
< Connection: keep-alive
< Proxy-Connection: keep-alive
<
* Ignore 1274 bytes of response-body
* TUNNEL_STATE switched to: 0
* Establish HTTP proxy tunnel to github.com:443
* Proxy auth using NTLM with user ''
> CONNECT github.com:443 HTTP/1.1
> Host: github.com:443
> Proxy-Authorization: NTLM TlRMTVNTUAADAAAAGAAYAHoAAAAYABgAkgAAABIAEgBIAAAADAAM
AFoAAAAUABQAZgAAAAAAAACqAAAABYKIogUBKAoAAAAPRABPAE0AQQBJAE4AMAAwADcAawBvAHMAdABp
AHgAUABSAE8ARwBSAEEATQBNAEUAUgDh1JtkpsObQAAAAAAAAAAAAAAAAAAAAADoWOvphE2qJZbbaDyf
GGzVbtbx1kWGEEo=
> User-Agent: curl/7.43.0
> Proxy-Connection: Keep-Alive
>
< HTTP/1.0 200 Connection established
<
* Proxy replied OK to CONNECT request
* ALPN, offering http/1.1
* Cipher selection: ALL:!EXPORT:!EXPORT40:!EXPORT56:!aNULL:!LOW:!RC4:@STRENGTH
* successfully set certificate verify locations:
*   CAfile: C:/Program Files/Git/mingw32/ssl/certs/ca-bundle.crt
  CApath: none
* TLSv1.2 (OUT), TLS header, Certificate Status (22):
* TLSv1.2 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Client hello (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS change cipher, Client hello (1):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*        subject: businessCategory=Private Organization; jurisdictionC=US; juris
dictionST=Delaware; serialNumber=5157550; street=548 4th Street; postalCode=9410
7; C=US; ST=California; L=San Francisco; O=GitHub, Inc.; CN=github.com
*        start date: 2014-04-08 00:00:00 GMT
*        expire date: 2016-04-12 12:00:00 GMT
*        subjectAltName: github.com matched
*        issuer: C=US; O=DigiCert Inc; OU=www.digicert.com; CN=DigiCert SHA2 Ext
ended Validation Server CA
*        SSL certificate verify ok.
> GET /nunit/nunit.git/info/refs?service=git-upload-pack HTTP/1.1
> Host: github.com
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: GitHub Babel 2.0
< Content-Type: application/x-git-upload-pack-advertisement
< Transfer-Encoding: chunked
< Expires: Fri, 01 Jan 1980 00:00:00 GMT
< Pragma: no-cache
< Cache-Control: no-cache, max-age=0, must-revalidate
< Vary: Accept-Encoding
<
001e# service=git-upload-pack
000000f42cde185d13a7abedb14910a5d2ee042a608c44af HEAD multi_ack thin-pack side-b
and side-band-64k ofs-delta shallow no-progress include-tag multi_ack_detailed n
o-done symref=HEAD:refs/heads/master agent=git/2:2.4.5~vmg-daemon-children-984-g
ec12e49
004fdc38d9e559dde703d2e9b879bb35481adfee54ab refs/heads/feature-sharedprojects
004cf0a019e49bd092a66880f8e9ddfb65cba4ade145 refs/heads/generic-constraints
````

So, my next couple of thoughts:

* `libCURL` shipped with GfW is built with support for `NTLM`, `Kerberos` and `SPNEGO` so it should actually provide you SSO.

    So try dropping that username+password bits from your proxy spec.

* If it does not work this way (or you want explicit credentials), make sure your username is OK.

    If you're in a Windows domain, your username has do be `DOMAIN\user`, and may require escaping that `\` by using `\\`. So I'd try that.

**Comment by Andrey-Bushman** [comment]
> When I used version 1.9.5 from here: https://git-scm.com/download/win it worked fine. :(((

It worked for me in earlier monthes (with my proxy settings in .gitconfig file as I wrote above). To be convinced of it once again I uninstall it http://git-for-windows.github.io/ and install this https://git-scm.com/download/win again. But now it doesn't work too... 

This is my current output for 1.9.5:

Developer@BUSHCOMP /d/_git.sandbox
$ git clone https://github.com/nunit/nunit.git
Cloning into 'nunit'...
fatal: unable to access 'https://github.com/nunit/nunit.git/': Unknown SSL proto
col error in connection to github.com:443

>  is this possible to negotiate with your admins to let direct connections to github's 443 for some time?
Admin told me he doesn't lock 443 port.

Developer@BUSHCOMP /d/_git.sandbox
$ openssl s_client -host github.com -port 443
Loading 'screen' into random state - done
CONNECTED(00000634)
5872:error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol:.\ssl\s
23_clnt.c:604:

>If you're in a Windows domain, your username has do be DOMAIN\user, and may require escaping that \ by using \\. So I'd try that.

But it worked earlier without domain name... Ok, I tried such variant:

[http]
    proxy = http://HYPROSTROY\\Developer:my_password@proxy2:8080

[https]
    proxy = https://HYPROSTROY\\Developer:my_password@proxy2:8080

But it didn't help me. I get the same problem.

**Comment by Andrey-Bushman** [comment]
Oh... I am inattentive donkey... Every 30 days we change our passwords. I changed it for the https settings in my .gitconfig file, but didn't the same for the http. Now I fixed it and all works fine. Forgive me guys that I spent your time. I appologize....

**Comment by Andrey-Bushman** [comment]
How can I remove this unuseful "issue"?

**Comment by kostix** [comment]
Andrey, are you sure you don't want to allow Git use single sign-on for you as I outlined in my previous comment?  That would solve your issue once and for all -- that is, `libCURL` would use `secur32.dll` of your Windows to authenticate you against the proxy.

**Comment by Andrey-Bushman** [comment]
kostix, how can I use it? Must I download `libCURL` from the http://curl.haxx.se/download.html and install only, or it is necessary to do additional settings?

**Comment by kostix** [comment]
Andrey, have you actually read this thread?

Git uses `libCURL` to use HTTP[S], so it's already bundled. `curl` is a command-line client for `libCURL` (bundled with Git as well).  `libCURL` shipped in GfW is built with support for all standard authentication methods providing SSO in Windows: Kerberos, NTLM and SPNEGO. All these auth mechanisms merely mediate between the server (proxy) and the system (your local Windows) when authenticating so basically they make Windows authenticate you using your system credentials.

TL;DR
Try dropping `user:password` part from your proxy definitions in Git condig and see if it works.

**Comment by Andrey-Bushman** [comment]
>Try dropping user:password part from your proxy definitions in Git config and see if it works.

I tried such wariants in my `.gitconfig` file, but it doesn't work:

proxy = proxy2:8080
proxy = http://@proxy2:8080
proxy = http://proxy2:8080

**Comment by Andrey-Bushman** [comment]
Oh, it doesn't work for Git for Windows 1.9.5, but it works fine for Git for Windows Git-2.4.5.1-4th-release-candidate:

proxy = http://@proxy2:8080
proxy = https://@proxy2:8080

THank you, kostix!

**Comment by kostix** [comment]
The `@` is not needed -- it merely separates the user+password part from the host part (it literally reads as "at").

Glad it helped.

**Comment by Andrey-Bushman** [comment]
No, it doesn't work without `@` char. I tried it.

**Comment by Andrey-Bushman** [comment]
kostix, can you see my question here: http://stackoverflow.com/questions/31343200/git-can-i-use-additional-some-common-config-file-for-all-users-of-our-company
if you have a time.
Thank you.

**Comment by kostix** [comment]
Oh, that's a very interesting find indeed! The `curl` library *requires* the user to pass it the `--proxy-user :` command-line option when they want to authenticate at the proxy using SSO. That ":" means "user:password" with *empty* user and password fields.

Hence the "@" character in the proxy spec suggests this way we select empty user and password *but* require authentication at the proxy server, while absent "@" supposedly means we expect no authentication request from the proxy at all!

That's a valuable information indeed!

Regarding your SO question: please don't abuse the bug tracker: this information is entirely irrelevant to this thread.
